File: curso_c_cpp/cpp/06_herencia_polimorfismo/ejemplos/figuras.cpp
/*
 * Clase abstracta (virtual pura) como interfaz: toda Figura tiene area().
 */
#include <iostream>
using namespace std;

class Figura {
public:
    virtual double area() = 0;       // virtual pura -> Figura es abstracta
    virtual ~Figura() {}
};

class Circulo : public Figura {
    double r;
public:
    Circulo(double radio) : r(radio) {}
    double area() override { return 3.14159265 * r * r; }
};

class Rectangulo : public Figura {
    double base, altura;
public:
    Rectangulo(double b, double h) : base(b), altura(h) {}
    double area() override { return base * altura; }
};

int main() {
    Figura *figuras[2];
    figuras[0] = new Circulo(2);
    figuras[1] = new Rectangulo(3, 4);

    for (int i = 0; i < 2; i++) {
        cout << "Area: " << figuras[i]->area() << endl;   // 12.566, 12
    }
    for (int i = 0; i < 2; i++) delete figuras[i];
    return 0;
}

File: curso_java/12_herencia_polimorfismo/ejemplos/Figuras.java
/*
 * Clase abstracta como plantilla: toda Figura debe implementar area().
 */
abstract class Figura {
    abstract double area();          // metodo abstracto (sin cuerpo)
}

class Circulo extends Figura {
    double r;
    public Circulo(double r) { this.r = r; }
    @Override
    double area() { return 3.14159265 * r * r; }
}

class Rectangulo extends Figura {
    double base, altura;
    public Rectangulo(double base, double altura) { this.base = base; this.altura = altura; }
    @Override
    double area() { return base * altura; }
}

public class Figuras {
    public static void main(String[] args) {
        Figura[] figuras = {
            new Circulo(2),
            new Rectangulo(3, 4)
        };
        for (Figura f : figuras) {
            System.out.printf("Area: %.4f%n", f.area());   // 12.566..., 12
        }
    }
}
